modules/core/arrow-annotations-processor/src/main/java/arrow/generic/CoproductFileGenerator.kt

```diff
@@ -0,0 +1,201 @@
+package arrow.generic
+
+import com.squareup.kotlinpoet.ClassName
+import com.squareup.kotlinpoet.CodeBlock
+import com.squareup.kotlinpoet.FileSpec
+import com.squareup.kotlinpoet.FunSpec
+import com.squareup.kotlinpoet.KModifier
+import com.squareup.kotlinpoet.ParameterSpec
+import com.squareup.kotlinpoet.ParameterizedTypeName
+import com.squareup.kotlinpoet.TypeAliasSpec
+import com.squareup.kotlinpoet.TypeVariableName
+
+import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
+
+import java.io.File
+
+private const val alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+private const val packageName = "arrow.generic"
+
+fun generateCoproducts(destination: File) {
```

Comment:
I also feel like I'm doing something wrong with the generics here too:

```kotlin
            val coproduct2: Coproduct2<Long, String> = 6L.cop<Long, String>()

            coproduct2.map<String, String, Int> { it.length } shouldBe 6L.cop<Long, Int>()
```
This appears to be valid in the editor, but the `A` type is incorrect for the receiver, I'm not sure if there's a way to deal with that? I feel like I'm doing something wrong here lol

Reply:
- As long as you move `KotlinPoet` to be `compileOnly`, I don't see any strong reason to not use it if it makes our lives easier on codegen.
- About how many coproducts needed, I believe 26 (number of characters on the alphabet) is more than enough to start with.
- The `6L.cop<Long, String>()` syntax is a bit hard to understand at the beginning, I think it's just "casting" 6L to be a `Coproduct<Long, String>` (which would be `Long` for this case), is it? I'm not sure whether there's a better approach for these castings honestly, so just a comment.

Reply:
Generating up to 22 is fine, that is what we do for the rest at this time. We don't use Kotlin Poet and do the code the generation manually as in the other processors to avoid introducing extra deps at compile time that users may already have. Having said that I have no strong opinions but would be nice to be consistent with the rest and avoid parts of code written in different styles since we have no plans to migrate all of our processors to KotlinPoet.

Reply:
There is no biased side in Coproduct and there should be no right biased map or other biased methods. A map over this Coproduct would have to contemplate all cases so it's esentially a specialization of fold that preserves the container. For example: https://github.com/milessabin/shapeless/wiki/Feature-overview:-shapeless-2.0.0#coproducts-and-discriminated-unions The poly function there contemplates all cases in which the Coproduct may find itself.